tck/src/main/java/org/eclipse/microprofile/config/tck/ConfigAccessorTest.java

```diff
@@ -0,0 +1,259 @@
+import org.jboss.arquillian.testng.Arquillian;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+/**
+ * @author <a href="mailto:struberg@apache.org">Mark Struberg</a>
+ */
+public class ConfigAccessorTest extends Arquillian {
+
+    private @Inject Config config;
+
+    @Deployment
+    public static WebArchive deploy() {
+        JavaArchive testJar = ShrinkWrap
+            .create(JavaArchive.class, "configValueTest.jar")
+            .addPackage(AbstractTest.class.getPackage())
+            .addClass(ConfigAccessorTest.class)
+            .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml")
+            .addAsServiceProvider(ConfigSource.class, ConfigurableConfigSource.class)
+            .as(JavaArchive.class);
+
+        AbstractTest.addFile(testJar, "META-INF/javaconfig.properties");
+
+        WebArchive war = ShrinkWrap
+            .create(WebArchive.class, "configValueTest.war")
+            .addAsLibrary(testJar);
+        return war;
+    }
+
+
+    @Test
+    public void testGetValue() {
+        Assert.assertEquals(config.access("tck.config.test.javaconfig.configvalue.key1", String.class).build().getValue(), "value1");
+    }
+
+    @Test
+    public void testGetValueWithDefault() {
+        ConfigAccessor<Integer> cfga = config.access("tck.config.test.javaconfig.configvalue.withdefault.notexisting", Integer.class)
+                  .withDefault(Integer.valueOf(1234)).build();
+
+        Assert.assertEquals(cfga.getValue(), Integer.valueOf(1234));
+    }
+
+    @Test
+    public void testGetValueWithStringDefault() {
+        ConfigAccessor<Integer> cfga = config.access("tck.config.test.javaconfig.configvalue.withdefault.notexisting", Integer.class)
+                .withStringDefault("1234").build();
+
+        Assert.assertEquals(cfga.getValue(), Integer.valueOf(1234));
+    }
+
+    /**
+     * Checks whether variable substitution works.
+     * The following situation is configured in javaconfig.properties:
+     * <pre>
+     * tck.config.variable.baseHost = some.host.name
+     * tck.config.variable.firstEndpoint = http://${tck.config.variable.baseHost}/endpointOne
+     * tck.config.variable.secondEndpoint = http://${tck.config.variable.baseHost}/endpointTwo
+     * </pre>
+     */
+    @Test
+    public void testVariableReplacement() {
+        Assert.assertEquals(config.access("tck.config.variable.firstEndpoint", String.class).build().getValue(),
+                "http://some.host.name/endpointOne");
+
+        Assert.assertEquals(config.access("tck.config.variable.secondEndpoint", String.class).build().getValue(),
+                "http://some.host.name/endpointTwo");
+
+        // variables in Config.getValue and getOptionalValue do not get evaluated otoh
```

Comment:
why are they not evaluated? for backwards compatibility?

Does it support nested variable replacement? e.g.  `${foo.${bar}}`
Does it support default value? e.g. `${foo:123}` -> returns `123` if the prop `foo` is not found
Does it support escaping? (so that the string `${` can appear without being an expression

Reply:
They are not evaluated for backward compatibility. The variables were evaluated only via configaccess object.
${foo.${bar}} should be evaluated recursively.
At the moment, it does not support default value if a particular variable does not exist. I suggest For the case of ${foo}, if `foo` does not exist, we could either fail with exception or not to evaluate the variable.
Good point of the escaping aspect, we have not talked about this. I think it is a good idea to introduce escaping... how about \${... escape a escape character will be \\${-> \${, ${ will be still escaped

Reply:
Hi, I'm an outsider here but just wanted to comment that it would be nice if there was a way to be able to use variable expansion directly on the `Config` interface.  Couldn't this be something set on `org.eclipse.microprofile.config.spi.ConfigBuilder`?